service/storage_proxy.cc

```diff
@@ -1750,6 +1750,31 @@ storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::tok
     std::vector<gms::inet_address> natural_endpoints = rs.get_natural_endpoints(token);
     std::vector<gms::inet_address> pending_endpoints = _token_metadata.pending_endpoints_for(token, keyspace_name);
 
+    if (get_local_storage_service().get_token_metadata().is_any_node_being_replaced() &&
+        rs.get_type() != locator::replication_strategy_type::local) {
+        // When a new node is started to replace an existing dead node, we want
+        // to make the replacing node take writes but do not count it for
+        // consistency level, because the replacing node can die and go away.
+        // To do this, we filter out the existing node being replaced from
+        // natural_endpoints and make the replacing node in the pending_endpoints.
+        //
+        // However, we can not apply the filter for LocalStrategy because
+        // LocalStrategy always has RF = 1 and returns the node itself as the
+        // natural_endpoints. If the replacing node had the same ip address as
+        // the node being replaced, we can not filter out the local node,
+        // otherwise all the query to keyspace with LocalStrategy will fail,
+        // e.g., the system keyspace.
+        auto original_natural_endpoints = natural_endpoints;
+        auto it = boost::range::remove_if(natural_endpoints, [] (gms::inet_address& p) {
+            return get_local_storage_service().get_token_metadata().is_being_replaced(p);
+        });
+        natural_endpoints.erase(it, natural_endpoints.end());
+        if (original_natural_endpoints.size() != natural_endpoints.size()) {
+            slogger.trace("creating write handler for token: {}, ks={}, original_natrual={}, natural={}, pending={}",
+                    token, keyspace_name, original_natural_endpoints, natural_endpoints, pending_endpoints);
```

Comment:
Why log twice here and below? But more importantly I do not understand what is going on here. Weren't node that is been replaced should have been filtered on gossiper level? get_natural_endpoints  does not suppose to return it.

Reply:
I guess you forget what we have discussed: https://github.com/scylladb/scylla/issues/5482#issuecomment-576165303

Reply:
It is mainly used for debug for this early version. I can drop it.

Reply:
I did not read it as literal. What is the point to do filtering in sp each time get_natural_endpoints is called? Do it once in 
get_natural_endpoints. Otherwise you will have to do it for all its callers. What about view building code for instance? Did you handle it here?

Reply:
I think it is wrong to filter inside get_natural_endpoints. What about the other callers of get_natural_endpoints that does not in the proxy write path. For example, the api in  api/api-doc/storage_service.json, repair to get neighbour in repair/repair.cc. 

The dead node to be replaced is still part of the ring. It should be in the get_natural_endpoints until we finish the replacing operation.

Reply:
Can you elaborate what should be done for the view building code? Is it in proxy service/storage_proxy.cc?

Reply:
Why would you want to get a dead and never comming back node in repair/repair.cc (can we even run repair in parallel with replace?)? Same with API. Why not return replacing node instead since it is the range owned now. Without knowing what the API is used for it is hard to tell what is best there.

And view building is the same as a write path. Just cl=one write. So I do not see why it is not suppose to be handled the same.

Reply:
Nothing says the node will never come back. The user can abort the replace operation and bring back the dead node back. 

The API is here:
```
 472          "path":"/storage_service/natural_endpoints/{keyspace}",
 473          "operations":[
 474             {
 475                "method":"GET",
 476                "summary":"This method returns the N endpoints that are responsible for storing the specified key i.e for replication. the endpoint responsible for this key",
 477                "type":"array",
 478                "items":{
 479                   "type":"string"
 480                },
```

Is the dead node still part of the ring when replacing? Yes. 
Should we tell user that the node being replaced is not part of the ring? No.

Can you point to me the relevant code for view build code?

Reply:
So report the old ring topology here by using some other function or introduce some other function to use everywhere else, though I would say that while replacing the dead node is not longer responsible for a range since it is dead. What about reapair code? Can it even run during replacement?

For view related code search for all uses of get natural and pending endpoints and look at those in view related files. I see some in `db/view/view.cc`. And if all writing code has to do the same than it has to be done in one place.

Reply:
sorage_proxy..c is the only place needs this filtering out replacing node thing. I think it is better to introduce a new function like get_natural_endpoints_without_node_being_replaced. Use this new function in  storage_proxy. 

We do not reject repair during replace operation. Two cases: 1) the replacing node is using the same IP address, the repair would repair with the node being replaced, should be ok 2) if the replacing node is using a different IP, the repair would repair with the old dead node, the repair with ignore the dead node.

Reply:
> Did you look into the use of pending endpoint in the view code?

I haven't.  I wanted to settle down a version without considering MV first.

> About repair and replacing simultaneously in the first case two different streaming operation will run. Is it even supported? Tested? Also it will stream the same data, so even if it will work this is not something that should be done.

We do not have this test in dtest or sct. Yes, I agree the repair is not helpfu in this case.

Reply:
> > Did you look into the use of pending endpoint in the view code?
> 
> I haven't. I wanted to settle down a version without considering MV first.
> 
How can you say that storage_proxy is the only place that needs it than?

Reply:
> sorage_proxy..c is the only place needs this filtering out replacing node thing. I think it is better to introduce a new function like get_natural_endpoints_without_node_being_replaced. Use this new function in storage_proxy.
> 
> We do not reject repair during replace operation. Two cases: 1) the replacing node is using the same IP address, the repair would repair with the node being replaced, should be ok 2) if the replacing node is using a different IP, the repair would repair with the old dead node, the repair with ignore the dead node.

Repair should not ignore the dead node. If it cannot repair every node then it should fail. In both cases above it should fail. In case 1, the repair should not succeed because replacing may be aborted and the old replica set is not repaired.

Reply:
The repair will fail because it is a partial repair, some of the nodes are down. Repair offers best effort, i.e. continue to repair even if some of the repair followers are down. In the end, it will report the repair failed.